src/nomadic/realtime/watchers.py

```diff
@@ -41,29 +61,113 @@ def _check_fastqs(self):
         )
 
         # Are any unprocessed?
-        self.unprocessed_fastqs = observed_fastqs.difference(self.processed_fastqs)
+        return observed_fastqs.difference(self.processed_fastqs)
 
-    def _all_fastqs_processed(self):
+    def catch_up_from_work_log(self) -> tuple[int, int]:
         """
-        Set all FASTQ files to processed
+        Read the log file and set all FASTQs that have been processed before
         """
-
-        self.processed_fastqs.update(self.unprocessed_fastqs)
+        if not os.path.exists(self.work_log_path):
+            return 0, 0
+
+        started_and_not_ended = dict()
+        with open(self.work_log_path, "r") as log_file:
+            number_fully_processed_fastqs = 0
+            for line in log_file:
+                # TODO verify that files match via hash
+                incr_id, action, fastqs = parse_log_line(line.strip())
+                # Filename in log is just the basename, so we need to prepend the directory
+                fastqs = [
+                    f"{self.barcode_fastq_dir}/{fastq}"
+                    for fastq in fastqs
+                    if self._is_fastq(fastq)
+                ]
+                if action is Action.START:
+                    # Multiple starts for the same increment ID are ok and will be just overriden, e.g. if the application is restarded
+                    started_and_not_ended[incr_id] = IncrementInfo(incr_id, fastqs)
+                elif action is Action.END:
+                    if incr_id not in started_and_not_ended:
+                        raise RuntimeError(
+                            f"Found END for step {incr_id} without START"
+                        )
+                    del started_and_not_ended[incr_id]
+                    number_fully_processed_fastqs += len(fastqs)
+                    self.processed_fastqs.update(fastqs)
+                else:
+                    raise RuntimeError(f"Unknown action in log: {action}")
+
+        number_reprocessed_fastqs = 0
+        for step_info in started_and_not_ended.values():
```

Comment:
Why "step_info" rather than "incr_info" if this is an IncrementInfo object?

Reply:
was a left over when I renamed, done